wandb/sdk/launch/sweeps/scheduler_optuna.py

```diff
@@ -0,0 +1,352 @@
+                log = "log" in param
+                distributions[param] = IntDistribution(
+                    extras.get("min"), extras.get("max"), log=log
+                )
+            else:
+                print(f"{LOG_PREFIX} Unknown parameter type, help! {param=}, {extras=}")
+
+        # Do we want to set the metric here? Note for the users, we access the metric
+        # when pruning, so we need to know what its called. defaults to "loss"
+        # epochs defaults to epoch, could also name step...
+
+        return distributions
+
+    def _load_db(self, sweep_name: str = "example-study"):
+        """
+        Create an optuna study with a sqlite backened for loose state management
+        """
+        # TODO(gst): add to validate function to confirm this exists, warn user
+        params = self._sweep_config.get("parameters", {})
+        epochs = params.get("epochs", {}).get("value") or 20  # could also use STEP here
+        pruner_args = params.get("pruner", {})
+
+        pruner = optuna.pruners.HyperbandPruner(
+            min_resource=pruner_args.get(
+                "min_resource", 1
+            ),  # TODO(gst): defaults are from tutorial
+            max_resource=epochs,
+            reduction_factor=pruner_args.get("reduction_factor", 3),
+        )
+
+        storage_name = f"sqlite:///{sweep_name}.db"
+        return optuna.create_study(
+            study_name=sweep_name,
+            storage=storage_name,
+            pruner=pruner,
+            load_if_exists=True,
+        )
+
+    def _start(self) -> None:
+        for worker_id in range(self._num_workers):
+            print(f"{LOG_PREFIX}Starting AgentHeartbeat worker {worker_id}\n")
+            agent_config = self._api.register_agent(
+                f"{socket.gethostname()}-{worker_id}",  # host
+                sweep_id=self._sweep_id,
+                project_name=self._project,
+                entity=self._entity,
+            )
+            self._workers[worker_id] = _Worker(
+                agent_config=agent_config,
+                agent_id=agent_config["id"],
+            )
+
+    def _heartbeat(self, worker_id: int) -> None:
+        # Make sure Scheduler is alive
+        if not self.is_alive():
+            return
+
+        # AgentHeartbeat wants a Dict of runs which are running or queued
+        _run_states: Dict[str, bool] = {}
+        for run_id, run in self._yield_runs():
+            # Filter out runs that are from a different worker thread
+            if run.worker_id == worker_id and run.state == SimpleRunState.ALIVE:
+                _run_states[run_id] = True
+
+        print(f"{LOG_PREFIX}OptunaHeartbeat sending: \n")
+        commands: List[Dict[str, Any]] = self._optuna_heartbeat(
+            self._workers[worker_id].agent_id,  # agent_id: str
+            _run_states,  # run_states: dict
+        )
+        print(f"{LOG_PREFIX}OptunaHeartbeat received {len(commands)} commands: \n")
+
+        for command in commands:
```

Comment:
hmmm, I vote we don't do any of this, instead we make the sweep schedule a regular W&B run and make the user handle it through the W&B runs UX.